[MRG] Fixed issue #4268 (bug in BernoulliNB).

hannawallach opened the pull request:
Fixed issue #4268 as discussed.

hannawallach committed:
Fixed issue #4268 (bug in BernoulliNB).

ogrisel commented:
Please include the test and change the title of this PR from `[WIP]` to `[MRG]` once done.

amueller commented:
Thanks @hannawallach :) Glad you find time for it!

hannawallach committed:
Added test for fix to issue #4268 (bug in BernoulliNB).

amueller commented:
LGTM, thanks :) Maybe add a comment that you are testing against the IR book formula, but it is fine the way it is.

hannawallach committed:
Added a comment explaining the test for the fix to issue #4268.

hannawallach commented:
Oh! Sure. Added a comment.

ogrisel commented on sklearn/tests/test_naive_bayes.py:
```diff
@@ -373,3 +373,24 @@ def test_check_accuracy_on_digits():
+    denom = np.tile(np.log(clf.class_count_ + 2.0), (X.shape[1], 1)).T
+
+    # Check manual estimate matches
+    assert_array_equal(clf.feature_log_prob_, (num - denom))
```
Actually, instead of or as a complement to this test, could you please write a test with the toy data of [table 13.1](http://nlp.stanford.edu/IR-book/html/htmledition/naive-bayes-text-classification-1.html) and reproduce the probabilities computed of [at the bottom of this page](http://nlp.stanford.edu/IR-book/html/htmledition/the-bernoulli-model-1.html#tab:nbtoy)?

hannawallach replied:
I can add it as a complement. That example won't highlight whether the original bug is fixed -- the bug won't appear if the number of classes is equal to 2 * alpha (which it is in that example).

ogrisel replied:
Well you can check that the `clf.feature_log_prob_` match the log of the conditional probabilities reported in:

http://nlp.stanford.edu/IR-book/html/htmledition/the-bernoulli-model-1.html#tab:nbtoy

WDYT?

hannawallach replied:
Seems reasonable. Just added such a test. LMK if you want me to change anything.

ogrisel commented on sklearn/tests/test_naive_bayes.py:
```diff
@@ -373,3 +373,27 @@ def test_check_accuracy_on_digits():
+
+    Tests that the feature log prob value computed by BernoulliNB when
+    alpha=1.0 is equal to the expression given in Manning, Raghavan,
+    and Schuetze's "Introduction to Information Retrieval" book.
```
Please include a the URL of the page of the Bernoulli model.

hannawallach committed:
Added link to IR book.

hannawallach committed:
Added test that checks that BernoulliNB gives the same probabilities
(class prior, feature probabilities, predictive probabilities) as the
example in Manning, Raghavan, and Schuetze's IR book
(http://nlp.stanford.edu/IR-book/html/htmledition/the-bernoulli-model-1.html).

hannawallach committed:
Fixed typo in test_bnb() that caused the Travis CI build to fail.

ogrisel commented:
This is great @hannawallach, thanks you very much for you contribution. I will squash your last 2 commits, update to whats_new.rst to document the fix and merge.

ogrisel commented:
I squashed into 2 commits: one for the fix and one for the tests. Thanks again @hannawallach for this fix!

amueller commented:
Should we add a whatsnew?
